plugin/updater_oneapi/regression_obj_oneapi.cc

```diff
@@ -0,0 +1,180 @@
+          int idx = pid[0];
+          bst_float p = Loss::PredTransform(preds_acc[idx]);
+          bst_float w = weights_acc[idx];
+          bst_float label = labels_acc[idx];
+          if (label == 1.0f) {
+            w *= scale_pos_weight;
+          }
+          out_gpair_acc[idx] = GradientPair(Loss::FirstOrderGradient(p, label) * w,
+                                            Loss::SecondOrderGradient(p, label) * w);
+        });
+      }).wait();
+    } else {
+      qu_.submit([&](cl::sycl::handler& cgh) {
+        auto preds_acc = preds_buf.get_access<cl::sycl::access::mode::read>(cgh);
+        auto labels_acc = labels_buf.get_access<cl::sycl::access::mode::read>(cgh);
+        auto out_gpair_acc = out_gpair_buf.get_access<cl::sycl::access::mode::write>(cgh);
+        cgh.parallel_for<class GetGradientNoWeights>(cl::sycl::range<1>(ndata), [=](cl::sycl::id<1> pid) {
+          int idx = pid[0];
+          bst_float p = Loss::PredTransform(preds_acc[idx]);
+          bst_float w = 1.0f;
+          bst_float label = labels_acc[idx];
+          if (label == 1.0f) {
+            w *= scale_pos_weight;
+          }
+          out_gpair_acc[idx] = GradientPair(Loss::FirstOrderGradient(p, label) * w,
+                                            Loss::SecondOrderGradient(p, label) * w);
+        });
+      }).wait();
+    }
+  }
+
+ public:
+  const char* DefaultEvalMetric() const override {
+    return Loss::DefaultEvalMetric();
+  }
+
+  void PredTransform(HostDeviceVector<float> *io_preds) override {
+    size_t const ndata = io_preds->Size();
+
+    cl::sycl::buffer<bst_float, 1> io_preds_buf(io_preds->HostPointer(), io_preds->Size());
+
+    qu_.submit([&](cl::sycl::handler& cgh) {
+      auto io_preds_acc = io_preds_buf.get_access<cl::sycl::access::mode::read_write>(cgh);
+      cgh.parallel_for<class PredTransform>(cl::sycl::range<1>(ndata), [=](cl::sycl::id<1> pid) {
+        int idx = pid[0];
+        io_preds_acc[idx] = Loss::PredTransform(io_preds_acc[idx]);
+      });
+    }).wait();
+  }
+
+  float ProbToMargin(float base_score) const override {
+    return Loss::ProbToMargin(base_score);
+  }
+
+  void SaveConfig(Json* p_out) const override {
+    auto& out = *p_out;
+    out["name"] = String(Loss::Name());
+    out["reg_loss_param"] = ToJson(param_);
+  }
+
+  void LoadConfig(Json const& in) override {
+    FromJson(in["reg_loss_param"], &param_);
+  }
+
+ protected:
+  RegLossParamOneAPI param_;
+
+  cl::sycl::queue qu_;
+};
+
+// register the objective functions
+DMLC_REGISTER_PARAMETER(RegLossParamOneAPI);
+
+XGBOOST_REGISTER_OBJECTIVE(SquaredLossRegressionOneAPI, LinearSquareLossOneAPI::Name())
+.describe("Regression with squared error with DPC++ backend.")
+.set_body([]() { return new RegLossObjOneAPI<LinearSquareLossOneAPI>(); });
+
+// TODO: Find a way to dispatch names of DPC++ kernels with various template parameters of loss function

```

Comment:
If you want we can work it out together.  I don't mind some refactoring for internal code first.

Reply:
Current approach looks fine.

This is more of a limitation of a DPC++ compiler - kernel name should be some globally-visible class name, so I can't define it inside like Loss structure. The only way I found up to this point is to define empty classes for each kernel outside Loss structure (somewhere nearby) and pass it as an additional template parameters.

In case of objective function it will be two additional template parameters (one in case of merged kernel).

In general case number of kernels can be much larger and it will also require explicit specialization of all potential combinations of templated classes/functions and it will go messy very fast.

I will probably communicate internally with DPC++ compiler team to find better solution.

Reply:
Implemented multiple loss functions using template parameters for kernels name for now.